src/app/services/product.service.ts

```diff
@@ -0,0 +1,89 @@
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { DomSanitizer } from '@angular/platform-browser';
+import { Observable, of } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+
+import { environment } from '../../environments/environment';
+import { Product } from '../models'
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ProductService {
+
+  readonly imageUrl = 'https://fruitshoppe.firebaseapp.com/images'; // URL of product images folder
+  readonly apiRoute = '/products';  // API route to retrieve products
+  readonly apiUrl: string;
+
+  constructor(
+    private sanitizer: DomSanitizer,
+    private http: HttpClient
+  ) {
+    // based on whether to use mocks or the API server, build the full API URL
+    this.apiUrl = environment.useMockApi ? `${environment.localApi}${this.apiRoute}` :
+      `${environment.remoteApi}${this.apiRoute}`;
+  }
+
+  /**
+   * Gets an observable that emits a product from the catalog.
+   * @param id the identifier of the product
+   */
+  getProduct(id: number): Observable<Product | null> {
+    const url = `${this.apiUrl}/${id}`;
+
+    return this.http.get<Product>(url).pipe(
+      map(product => this.setImageUrl(product)),
+      catchError(this.handleError<Product>('getProduct', null))
+    );
+  }
+
+  /**
+   * Gets an observable that emits a list of products from the catalog.
+   * @param query An optional query string to search for products
+   */
+  getProducts(query?: string): Observable<Product[]> {
+    // searching for products will not work for mocks
+    const url = query ? `${this.apiUrl}?q=${query}` : this.apiUrl;
+
+    return this.http.get<Product[]>(url).pipe(
+      map(products => products.map(product => this.setImageUrl(product))),
+      catchError(this.handleError<Product[]>('getProducts', []))
+    );
+  }
+
+  /**
+   * Provides a selector to allow handling of errors from service calls.
+   * @param operation the operation display name that encountered an error
+   * @param result the return value that should be passed to the next operator
+   */
+  private handleError<T>(operation = 'operation', result?: T) {
+    return (error: any): Observable<T> => {
```

Comment:
I think you avoid using the `any` type here. TypeScript should have an `Error` type available to you:

```TypeScript
interface Error {
    name: string;
    message: string;
    stack?: string;
}
```

https://github.com/microsoft/TypeScript/blob/master/lib/lib.es5.d.ts#L972

Reply:
added